### ISSUE TITLE
parallel flow_mpi crashes when thresholdPressures are used

### ISSUE THREAD

**Issue by blattms** [OPM/opm-autodiff#589]
Currently flow_mpi does not run for Norne as it uses the threshold pressure of the whole grid (as specified in the eclipse file) while the grid in the simulation only resembles a part of the grid. This crashes with
```
Program threw an exception: [/home/mblatt/DUNE-test/opm-autodiff/opm/autodiff/BlackoilModelBase_impl.hpp:383] Illegal size of threshold_pressures input ( 153924 ), must be equal to number of faces + nncs ( 78316 + 0 ).
```

The problem might have been produced when refactoring `flow.cpp` and `flow_mpi.cpp`. Currently the threshold pressures are computed within `FlowMainBase::setupState` which works on the whole grid.  Later the grid and the data is distributed during `distributeData`, which does not and never did distribute the threshold pressures.

As it worked somewhen my guess is that previously the thresholdPressures were setup after loadbalancing the grid and the data. If that is still possible I would love to resort to this behaviour:
```
setupSate();
distributeData();
setupLocalState(); // read threshold pressure with the distributed grid.
```

To me it seems like all the functions (`thresholdPressures`, `thresholdPressuresNNC`, and `computeMaxDp`) should work if the grid resembles only a part of the one defined in the eclipse file. As I might be missing something, I would appreciate it if someone would confirm this few.

**Comment by andlaus** [comment]
> To me it seems like all the functions (thresholdPressures, thresholdPressuresNNC, and computeMaxDp) should work if the grid resembles only a part of the one defined in the eclipse file. As I might be missing something, I would appreciate it if someone would confirm this few. 

I suppose you're missing something: if the threshold pressures need to be calculated from the initial condition (i.e., if some of them are defaulted), they will be wrong for some processes if some PVT region is distributed over more than a single process. (this can relatively easily be fixed with some additional MPI communication, though.)

**Comment by alfbr** [comment]
Was this fallout from the new pvt code? Any chance of getting a fast fix? Note to self, we really need Jenkins test for mpi version.

**Comment by atgeirr** [comment]
> The problem might have been produced when refactoring flow.cpp and flow_mpi.cpp

Probably true, but as @andlaus points out,

> if the threshold pressures need to be calculated from the initial condition (i.e., if some of them are defaulted), they will be wrong for some processes if some PVT region is distributed over more than a single process

Point is, we did not use to support calculating threshold pressures from initial condition (it's not needed for Norne), but now we do. I agree that the best is probably to communicate the threshold info.

**Comment by acelster** [comment]
Can or should these threshold pressures change over time for more
complicated data sets (than Norne) and/or are they mostly just
input/initial state dependent?

P.s. Yes, I am now monitoring these Github updates more carefully now that
Malik and I are working on the OPM use case for our CloudLightning EU
project.

BTW: I have also just announced two positions in my research group on
jobbnorge. Please send good candidates my way so that they also can help us
out on parallelizing OPM!

https://www.jobbnorge.no/ledige-stillinger/stilling/122196/postdoctoral-fellowship-in-high-performance-computing-hpc-related-to-the-eu-project-cloudlightning

https://www.jobbnorge.no/ledige-stillinger/stilling/122202/phd-fellowship-in-high-performance-computing-hpc

Anne C. Elster
from Austin, Texas

**Comment by dr-robertk** [comment]
@acelster: You want to parallelize it? Again? ;-) 
Help is definitely welcome.

**Comment by alfbr** [comment]
Threshold pressures can change over time (at least in principle). However, I am not aware of any usage of that. The current usage I have seen is to create skewed oil-water contacts.

Good luck with finding candidates, help is always welcome :-)

**Comment by andlaus** [comment]
> Was this fallout from the new pvt code? Any chance of getting a fast fix? Note to self, we really need Jenkins test for mpi version.

no. The PVT code is completely orthogonal (hopefully). if the threshold-pressure-from-initial-condition code is really the culprit, it broke sometime in December last year. (i.e., it looks like nobody used `flow_mpi` seriously in the mean time?)

**Comment by alfbr** [comment]
No, December was too busy. However, now a number of people depend on it, so we need to fix it fast.

**Comment by andlaus** [comment]
@acelster 
> Can or should these threshold pressures change over time for more complicated data sets (than Norne) and/or are they mostly just input/initial state dependent?

since threshold pressures are a *freaking* big hack in my opinion, I think they *should* not change during a simulation. (in fact, they *should* always be 0 ;).) Since this is ECL, I suppose they *can* be changed in the SCHEDULE section, but then the default values will (hopefully) not need to be recalculated and the changes required for the simulator code would be pretty limited.

**Comment by alfbr** [comment]
I will need to read up a bit in the documentation and current usage tomorrow. It is difficult to have an informed opinion now, but from the top of my head, I'd be willing to sacrifice a lot of threshold pressure functionality short term in the mpi version to get it working again.

**Comment by atgeirr** [comment]
Threshold pressures cannot change in the schedule, so it is not *that* ugly...

**Comment by alfbr** [comment]
> they will be wrong for some processes if some PVT region is distributed over more than a single process

@andlaus You probably mean equilibration region interfaces rather than PVT regions here.

>this can relatively easily be fixed with some additional MPI communication, though.

I agree. Nice catch with the defaulted values. Defaulted values are supposed to be replaced by the maximum potential pressure (for phases present in the cells) for the interface between the equilibration regions. If the interface is divided between regions, each region will need to communicate its maximum in order to establish the interface maximum. The potential pressures in turn is determined by the initialisation, so that will need to be done first.

>Threshold pressures cannot change in the schedule, so it is not that ugly...

In principle, this is an implementation choice (or a input file format choice if you prefer). From the simulator point of view you can change the pressures at any time (risking convergence issues by bringing the system out of equilibrium of course). However, there seems to be no wish for such a feature, so I believe we can safely assume that users will be content with threshold pressures set only initially.

**Comment by andlaus** [comment]
> Threshold pressures cannot change in the schedule, so it is not that ugly...

phew! so for the sequential case the current implementation should be comprehensive?

> In principle, this is an implementation choice (or an input file format choice if you prefer).

I think @atgeirr's statement was referring to what The Truth is (i.e., what Eclipse does ;).

> @andlaus You probably mean equilibration region interface rather than PVT region here.

right. I meant the kind of region to which's boundary the threshold pressures apply. I should have said XYZ-region ;)

**Comment by blattms** [comment]
So distributing the initial values during `distributeData` is the
method of choice. Should be easy enough for me ;)

*Sigh* It is a pitty that deer (and OPM) do not stay put if want to
take a shot. Would be much easier than a running target but a lot more
boring.
